File: KV_Assignment_Module11_Wordcloud.R
install.packages("rvest")
install.packages("XML")
install.packages("magrittr")


library(rvest)
library(XML)
library(magrittr)
library(xml2)

######### Apple watch Reviews ###########

aurl <-"https://www.amazon.in/Apple-Watch-GPS-44mm-Aluminium/dp/B07XWYJDDW/ref=sr_1_13?dchild=1&qid=1593212912&refinements=p_89%3AApple%2Cp_6%3AA14CZOWI0VEHLG%7CAT95IG9ONZD7S&s=electronics&sr=1-13#customerReviews"

reviews <- NULL

for (i in 1:20){
  mac <- read_html(as.character(paste(aurl,i,sep ="=")))
  rev <- mac %>% html_nodes(".review-text") %>% html_text()
  reviews <- c(reviews,rev)
}

write.table(reviews,"applewatch.txt")
getwd()

##################################

txt <- reviews

str(txt)
length(txt)

# Corpus
install.packages("tm")
library(tm)

x <- Corpus(VectorSource(txt))

inspect(x[1])
inspect(x[160])

x <- tm_map(x, function(x) iconv(enc2utf8(x), sub='byte'))
?tm_map

# Data Cleansing
x1 <- tm_map(x, tolower)
inspect(x1[1])

x1 <- tm_map(x1, removePunctuation)
inspect(x1[1])

inspect(x1[5])
x1 <- tm_map(x1, removeNumbers)
inspect(x1[1])

x1 <- tm_map(x1, removeWords, stopwords('english'))
inspect(x1[1])
inspect(x1[3])

# striping white spaces 
x1 <- tm_map(x1, stripWhitespace)
inspect(x1[1])

# Term document matrix 
# converting unstructured data to structured format using TDM

tdm <- TermDocumentMatrix(x1)
tdm
dtm <- t(tdm) # transpose
dtm <- DocumentTermMatrix(x1)

tdm <- as.matrix(tdm)
dim(tdm)

tdm[1:20, 1:20]

inspect(x[3])

# Bar plot
w <- rowSums(tdm)
w

w_sub <- subset(w, w >= 65)
w_sub

barplot(w_sub, las=1, col = rainbow(30))

# Term phone repeats maximum number of times
x1 <- tm_map(x1, removeWords, c('iphone','even','iphone','just','now','read','watch','apple','display','get','series'))
x1 <- tm_map(x1, stripWhitespace)

tdm <- TermDocumentMatrix(x1)
tdm

tdm <- as.matrix(tdm)
tdm[100:109, 1:20]

# Bar plot after removal of the term 'phone'
w <- rowSums(tdm)
w

w_sub <- subset(w, w >= 50)
w_sub
sort(w_sub)

barplot(w_sub, las=2, col = rainbow(30))

##### Word cloud #####
install.packages("wordcloud")
library(wordcloud)

wordcloud(words = names(w_sub), freq = w_sub)

w_sub1 <- sort(rowSums(tdm), decreasing = TRUE)
head(w_sub1)

wordcloud(words = names(w_sub1), freq = w_sub1) # all words are considered

# better visualization
wordcloud(words = names(w_sub1), freq = w_sub1, random.order=F,colors=rainbow(30),scale=c(2,0.5),rot.per=0.4)
windows()
wordcloud(words = names(w_sub1), freq = w_sub1, random.order=F,colors= rainbow(30),scale=c(3,0.5),rot.per=0.3)
?wordcloud

############# Wordcloud2 ###############
install.packages("wordcloud2")
library(wordcloud2)

w1 <- data.frame(names(w_sub), w_sub)
colnames(w1) <- c('word', 'freq')
?wordcloud2

wordcloud2(w1, size=0.4, shape='circle')

wordcloud2(w1, size=0.3, shape = 'triangle')

############

File: KV_Assignment_Module11_Wordcloud.py
# -*- coding: utf-8 -*-
"""
Created on Sat Jun 27 06:54:46 2020

@author: KavithaV
"""

import requests
from bs4 import BeautifulSoup  # for web scraping

import matplotlib.pyplot as plt
from wordcloud import WordCloud

#header={'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/77.0.3865.90 Safari/537.36'}

# Creating empty review list
applewatch_reviews = []

for i in range(1,30):
    phone = []
    url = "https://www.amazon.in/Apple-Watch-GPS-44mm-Aluminium/dp/B07XWYJDDW/ref=sr_1_13?dchild=1&qid=1593212912&refinements=p_89%3AApple%2Cp_6%3AA14CZOWI0VEHLG%7CAT95IG9ONZD7S&s=electronics&sr=1-13#customerReviews"+str(i)
    header={'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/77.0.3865.90 Safari/537.36'}
    response = requests.get(url,headers = header)
    # creating soup object to iterate over the extracted content
    #soup = bs(response.content, "html.parser")
    soup = BeautifulSoup(response.text,"lxml")
    # Extract the content under the specific tag
    reviews = soup.find_all("div",{"data-hook":"review-collapsed"})
    for i in range(len(reviews)):
        phone.append(reviews[i].text)
    # Adding the reviews of one page to empty list which in future contains all the reviews
    applewatch_reviews = applewatch_reviews + phone
    
# Writing reviews in a text file
with open('applewatch.txt','w', encoding = 'utf8') as output:
    output.write(str(applewatch_reviews))

# Joinining all the reviews into single paragraph 
ip_rev_string = " ".join(applewatch_reviews)

import nltk
import re
from nltk.corpus import stopwords
from nltk.stem.wordnet import WordNetLemmatizer
wordnet = WordNetLemmatizer()
nltk.download('wordnet')

# Removing unwanted symbols incase if exists
ip_rev_string = re.sub("[^A-Za-z" "]+"," ",ip_rev_string).lower()
ip_rev_string = re.sub("[0-9" "]+"," ",ip_rev_string)

# words that contained in iphone XR reviews
ip_reviews_words = ip_rev_string.split(" ")

#TFIDF
from sklearn.feature_extraction.text import TfidfVectorizer
vectorizer=TfidfVectorizer(ip_reviews_words,use_idf=True,ngram_range=(1, 3))
X=vectorizer.fit_transform(ip_reviews_words)

with open("D:/Data science/Module 11 - Text mining/Assignment/Assignment/stopwords.txt","r") as sw:
    stop_words = sw.read()

stop_words = stop_words.split("\n")

stop_words.extend(["watch","mobile","ios","apple","phone","amazon","good","xr","product","great","camera","price"])
ip_reviews_words = [w for w in ip_reviews_words if not w in stop_words]


# Joinining all the reviews into single paragraph 
ip_rev_string = " ".join(ip_reviews_words)

# WordCloud can be performed on the string inputs. Hence all the reviews are merged into single string
# Corpus word cloud

wordcloud_ip = WordCloud(
                      background_color='black',
                      width=1800,
                      height=1400
                     ).generate(ip_rev_string)
plt.figure(1)
plt.imshow(wordcloud_ip)

# positive words # Choose the path for +ve words stored in system
with open("D:/Data science/Module 11 - Text mining/Assignment/Assignment/positive-words.txt","r") as pos:
  poswords = pos.read().split("\n")
  

# Positive word cloud
# Choosing the only words which are present in positive words
ip_pos_in_pos = " ".join ([w for w in ip_reviews_words if w in poswords])
wordcloud_pos_in_pos = WordCloud(
                      background_color='White',
                      width=1800,
                      height=1400
                     ).generate(ip_pos_in_pos)

plt.figure(2)
plt.imshow(wordcloud_pos_in_pos)

  
# negative words  Choose path for -ve words stored in system
with open("D:/Data science/Module 11 - Text mining/Assignment/Assignment/negative-words.txt","r") as neg:
  negwords = neg.read().split("\n")

# negative word cloud
# Choosing the only words which are present in negwords
ip_neg_in_neg = " ".join ([w for w in ip_reviews_words if w in negwords])

wordcloud_neg_in_neg = WordCloud(
                      background_color='black',
                      width=1800,
                      height=1400
                     ).generate(ip_neg_in_neg)
plt.figure(3)
plt.imshow(wordcloud_neg_in_neg)


